src/librustc/mir/interpret/mod.rs

```diff
@@ -238,6 +242,127 @@ pub fn specialized_decode_alloc_id<
+            .take(data_offsets.len())
+            .collect();
+
+        AllocIdDecoder {
+            decoding_state: Mutex::new(decoding_state),
+            data_offsets,
+        }
+    }
+
+    pub fn decode_alloc_id<'a, 'tcx, D>(&self,
+                                        decoder: &mut D,
+                                        session_id: DecodingSessionId)
+                                        -> Result<AllocId, D::Error>
+        where D: TyDecoder<'a, 'tcx>,
+              'tcx: 'a,
+    {
+        // Read the index of the allocation
+        let idx = decoder.read_u32()? as usize;
+
+        // Check the decoding state, see if it's already decoded or if we should
+        // decode it here.
+        let alloc_id = loop {
+            let mut decoding_state = self.decoding_state.lock();
+            let entry = &mut decoding_state[idx];
+
+            match *entry {
+                State::Done(alloc_id) => {
+                    return Ok(alloc_id);
+                }
+                State::Empty => {
+                    // We are allowed to decode
+                    let alloc_id = decoder.tcx().alloc_map.lock().reserve();
+                    *entry = State::InProgress(session_id, alloc_id);
+                    break alloc_id;
+                }
+                State::InProgress(other_session_id, alloc_id) => {
+                    if other_session_id == session_id {
+                        // Don't recurse.
+                        return Ok(alloc_id);
+                    } else {
+                        // Just spin. We could also decode concurrently.
+                    }
+                }
+                State::Poisoned => {
+                    panic!("poisoned")
+                }
+            }
+        };
+
+        let pos = self.data_offsets[idx] as usize;
+
+        let alloc: Result<AllocType<_>, D::Error> = decoder.with_position(pos, |decoder| {
+            match AllocKind::decode(decoder)? {
+                AllocKind::Alloc => {
+                    let allocation = <&'tcx Allocation as Decodable>::decode(decoder)?;
+                    trace!("decoded alloc {:?} {:#?}", alloc_id, allocation);
+                    Ok(AllocType::Memory(allocation))
+                },
+                AllocKind::Fn => {
+                    trace!("creating fn alloc id");
+                    let instance = ty::Instance::decode(decoder)?;
+                    trace!("decoded fn alloc instance: {:?}", instance);
+                    Ok(AllocType::Function(instance))
+                },
+                AllocKind::Static => {
+                    trace!("creating extern static alloc id at");
+                    let did = DefId::decode(decoder)?;
+                    Ok(AllocType::Static(did))
+                }
+            }
+        });
+
+        match alloc {
+            Ok(alloc) => {
+                decoder.tcx().alloc_map.lock().set_id_alloc(alloc_id, alloc);
```

Comment:
This should use the `set_id_same_memory` function I introduced in my PR. `AllocKind::Fn` and `AllocKind::Static` should still use `create_fn_alloc` and `intern_static`.

Reply:
> This should use the set_id_same_memory function I introduced in my PR.

With the spinning solution (which again, I don't necessarily prefer) setting things in the `alloc_map` is not racy, so asserting that no key existed should be correct (and avoids doing a comparison).

> Can instance decoding potentially be expensive? Or DefId decoding?

It's not free. I think we should cache statics and fns too, if it doesn't complicate the algorithm.

Reply:
Yeah, you could just have a `set_id_memory` which just asserts that there was no existing entry.

Reply:
It does that already, if I'm not mistaken.

Reply:
Your accepts `AllocType`, which is incorrect, since `AllocType::Function` and `AllocType::Static` are interned. It should only accept allocations.

Reply:
I'll look into that. Thanks for the hint!